File: backend/app/services/interview_complete.py
"""
Single source of truth for marking an interview completed and closing
its WebSocket room with role-specific messages.
"""

from datetime import datetime
from typing import Optional, Tuple

from sqlalchemy.orm import Session

from app.models.interview import Interview

COMPLETION_MESSAGES = {
    "RECRUITER": {
        "recruiter": "Interview completed by recruiter.",
        "student": "Interview completed by recruiter.",
        "default": "Interview completed by recruiter.",
    },
    "TAB_SWITCH": {
        "recruiter": "Interview automatically completed due to candidate tab switching.",
        "student": "Interview terminated due to tab switching. The interview has been marked as completed.",
        "default": "Interview completed.",
    },
}


def completion_ws_payload(interview_id: str, reason: str, role: str) -> dict:
    messages = COMPLETION_MESSAGES.get(reason, COMPLETION_MESSAGES["RECRUITER"])
    message = messages.get(role) or messages.get("default")
    return {
        "type": "interview-completed",
        "reason": reason,
        "interview_id": str(interview_id),
        "message": message,
    }


async def complete_interview(
    db: Session,
    interview_id: str,
    reason: str = "RECRUITER",
) -> Tuple[Optional[Interview], bool]:
    """
    Mark interview completed and notify connected peers.

    Returns (interview, newly_completed).
    If already completed, returns (interview, False) without re-closing the room.
    """
    from app.api.signaling import manager

    interview = (
        db.query(Interview)
        .filter(Interview.id == interview_id)
        .first()
    )

    if not interview:
        return None, False

    if interview.status == "completed":
        return interview, False

    interview.status = "completed"
    interview.completed_at = datetime.utcnow()

    db.commit()
    db.refresh(interview)

    await manager.close_room(str(interview_id), reason=reason)

    return interview, True

File: backend-node/src/services/interviewComplete.js
import { prisma } from '../db/prisma.js'

export const COMPLETION_MESSAGES = {
  RECRUITER: {
    recruiter: 'Interview completed by recruiter.',
    student: 'Interview completed by recruiter.',
    default: 'Interview completed by recruiter.',
  },
  TAB_SWITCH: {
    recruiter: 'Interview automatically completed due to candidate tab switching.',
    student:
      'Interview terminated due to tab switching. The interview has been marked as completed.',
    default: 'Interview completed.',
  },
}

export function completionWsPayload(interviewId, reason, role) {
  const messages = COMPLETION_MESSAGES[reason] || COMPLETION_MESSAGES.RECRUITER
  const message = messages[role] || messages.default
  return {
    type: 'interview-completed',
    reason,
    interview_id: String(interviewId),
    message,
  }
}

export async function completeInterview(interviewId, reason = 'RECRUITER') {
  const interview = await prisma.interview.findUnique({ where: { id: interviewId } })
  if (!interview) return { interview: null, newlyCompleted: false }
  if (interview.status === 'completed') {
    return { interview, newlyCompleted: false }
  }

  const updated = await prisma.interview.update({
    where: { id: interviewId },
    data: {
      status: 'completed',
      completedAt: new Date(),
    },
  })

  const { manager } = await import('../ws/signaling.js')
  await manager.closeRoom(String(interviewId), reason)
  return { interview: updated, newlyCompleted: true }
}
